[browser] [wasm] Request Streaming upload

campersau opened the pull request:
Uses a `ReadableStream` as fetch request body and reads chunks from .NET `HttpContent` stream.
The feature needs to be enabled via `HttpRequestOptionsKey<bool>("WebAssemblyEnableStreamingRequest")`. A helper could be added to [`WebAssemblyHttpRequestMessageExtensions`](https://github.com/dotnet/aspnetcore/blob/9402bfac90a695bb732dff17ba624801076df77f/src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs#L134-L151).
It is only supported using HTTP/2 or higher and is currently only implemented in Chromium (Chrome / Edge) (see https://github.com/dotnet/runtime/issues/36634#issuecomment-1418701286).

~It currently tries to get the buffer size from the browser via [`desiredSize`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableByteStreamController/desiredSize) which in my testing was always `0`. We can also remove this detection and just use the fixed default buffer size.
I also enabled [`autoAllocateChunkSize`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/ReadableStream#autoallocatechunksize) which provides us a [`byobRequest`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableByteStreamController/byobRequest) so we can copy directly into the browsers request buffer. Unfortunately https://github.com/WebAssembly/design/issues/1162 is not yet implemented which means we can't do a zero-byte transfer by passing the `ArrayBuffer` to .NET and avoid any buffers in  .NET.~
The default buffer size is currently set to `65536`. Maybe it should also be made configurable via a `HttpRequestOptionsKey`?

Exceptions from reading the `HttpContent` stream will fail the request with the thrown exception.
The special case for `StringContent` is still preferred and bypasses streaming.

I added some `BrowserHttpHandler_*` tests to `ResponseStreamTest.cs` analog to the `WebAssemblyEnableStreamingResponse` one. Maybe these `BrowserHttpHandler_*` test should be moved in their own file?
I duplicated the `PostAsync_ThrowFromContentCopy_RequsetFails` tests to use remote server, as the `Http2LoopbackServer` does not work in the browser.

Do I need to adjust something regarding `FEATURE_WASM_THREADS`?

Fixes https://github.com/dotnet/runtime/issues/36634

campersau committed:
progress

campersau committed:
progress try out byob

campersau committed:
back to enqueue again

campersau committed:
finalize request stream test

campersau committed:
checkpoint

campersau committed:
reject fetch with read stream exception

campersau committed:
fix build

pavelsavara reviewed (commented):
@campersau thanks for doing this! I really appreciate it.

I left few initial comments. Overall it seems sound implementation to me.

Do we have test for both streaming request and streaming response on same call ?
Do we dispose all the JSObject we transmit ? 
Could the `pull` callback become static ? (so that it doesn't allocate for Delegate)

Regarding FEATURE_WASM_THREADS, I think you don't have to do anything about it and it would already work.
I plan to re-write it once more for MT, because I would like to dispatch the calls more asynchronously. Currently it's synchronously blocking the caller, which is not great. And to be able to do that I will have to do more shuffling with the way how pointers/buffers are pinned (on caller stack).
I don't know details yet. But overall it means I will do it for your new code as well.

campersau commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -41,14 +57,61 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+        const body = new ReadableStream({
+            type: "bytes",
+            autoAllocateChunkSize: 65536,
+            pull(controller: ReadableByteStreamController) {
```
> Maybe you need to change signature to Func<JSObject,int,Task> and await it here.

I would love to but the js import generator can't do that.
```
error SYSLIB1072: The type 'System.Func<System.Runtime.InteropServices.JavaScript.JSObject, int, System.Threading.Tasks.Task>' is not supported by source-generated JavaScript interop. The generated source will not handle marshalling of parameter 'pull'.
```
https://github.com/dotnet/runtime/blob/08b66559ed06d4d97fb09dfa0d20163f0865deef/src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSGeneratorFactory.cs#L270-L277

pavelsavara replied:
> > Maybe you need to change signature to Func<JSObject,int,Task> and await it here.
> I would love to but the js import generator can't do that.

That's a shame, sorry about that. I spent few hours today trying quick hack to support it, because I thought Task of void is not really generic type and it would be easy hack.

But it needs to be done properly. That means support for generic types with generic type arguments. In generator as well as in the runtime. There are quite few useful other types, like `Task<byte[]>` which would be good to have.

campersau committed:
use createPromiseController in pull

campersau committed:
dispose controller

campersau committed:
swap delegate allocation with state allocation

campersau committed:
add response / request streaming test

campersau commented:
> Do we have test for both streaming request and streaming response on same call ?

I added a test which does both request and response streaming in https://github.com/dotnet/runtime/pull/91295/commits/487ef337dfb31cbc6e5defe0ec069287edc99d62

> Do we dispose all the JSObject we transmit ?

JSObject controller now gets disposed in https://github.com/dotnet/runtime/pull/91295/commits/e2193e6c2db9e56872b2ed95a283809cdff12480

> Could the pull callback become static ? (so that it doesn't allocate for Delegate)

I made it static and moved all the state to a separate class which is send to JS and then back to .NET in https://github.com/dotnet/runtime/pull/91295/commits/92ae1327fc303cb7f7d2ade16a0925fc1903c95c

campersau committed:
use createPromiseController for fetch

campersau committed:
Revert "use createPromiseController for fetch"

This reverts commit 1ebb07db6122cf825426f42b50634ea1384effbc.

campersau commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -41,14 +57,61 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+}
+
+export function http_wasm_fetch_stream(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, pull: (controller: ReadableByteStreamControllerExtension, desired_size: number) => void): Promise<ResponseExtension> {
+    return new Promise((resolve_fetch, reject_fetch) => {
```
I tried to use it here as well but it was unreliable as it often crashed mono (`mono_exit(1, exception)` is getting called) but not always.

```
 fail: MONO_WASM:    at System.Net.Http.Functional.Tests.ResponseStreamTest.<>c__DisplayClass9_0.<BrowserHttpHandler_StreamingRequest_ThrowFromContentCopy_RequestFails>b__6(Task _)
           at System.Threading.Tasks.ContinuationResultTaskFromTask`1[[System.Int32, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
           at System.Threading.Tasks.Task.<>c.<.cctor>b__281_0(Object obj)
           at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
        --- End of stack trace from previous location ---
           at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
           at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
        --- End of stack trace from previous location ---
           at System.Net.Http.BrowserHttpHandler.ReadableStreamPull(JSObject controller, Int32 desiredSize, Object state)
        Error: One of the identified items was in an invalid format.
            at Yn (http://127.0.0.1:33075/_framework/dotnet.runtime.js:3:32022)
            at http://127.0.0.1:33075/_framework/dotnet.runtime.js:3:29829
            at http://127.0.0.1:33075/_framework/dotnet.runtime.js:3:179120
            at Tl (http://127.0.0.1:33075/_framework/dotnet.runtime.js:3:180290)
            at do_icall (http://127.0.0.1:33075/_framework/dotnet.native.wasm:wasm-function[318]:0x1f13d)
            at do_icall_wrapper (http://127.0.0.1:33075/_framework/dotnet.native.wasm:wasm-function[247]:0x1c1f2)
            at mono_interp_exec_method (http://127.0.0.1:33075/_framework/dotnet.native.wasm:wasm-function[240]:0xf2eb)
            at interp_entry (http://127.0.0.1:33075/_framework/dotnet.native.wasm:wasm-function[311]:0x1ebbc)
            at interp_entry_static_0 (http://127.0.0.1:33075/_framework/dotnet.native.wasm:wasm-function[346]:0x1fd2d)
            at wasm_native_to_interp_System_Private_CoreLib_ThreadPool_BackgroundJobHandler (http://127.0.0.1:33075/_framework/dotnet.native.wasm:wasm-function[219]:0xd162)
```

You can try out this commit https://github.com/dotnet/runtime/pull/91295/commits/1ebb07db6122cf825426f42b50634ea1384effbc and run this test a few times :
```
./dotnet.sh build /t:test /p:Outerloop=true /p:TargetOS=browser /p:TargetArchitecture=wasm /p:Configuration=Release src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj /p:xunitoptions="-method System.Net.Http.Functional.Tests.SocketsHttpHandler_ResponseStreamTest.BrowserHttpHandler_StreamingRequest_ThrowFromContentCopy_RequestFails"
```

Might be related to `async void`?

pavelsavara replied:
I will try to have look tomorrow, should I commit into this branch of have my own ?

campersau replied:
You can commit here as well.

pavelsavara commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -41,14 +57,64 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+                }
+                c.__resolve_pull = pull_promise_control.resolve;
+                c.__reject_fetch = reject_fetch;
+                pull(c, desired_size || 0, pull_state);
```
is this `pull` calling itself ?

pavelsavara replied:
I see that a parameter, could you please rename it so it doesn't clash ?

campersau replied:
No, it's calling the delegate which is passed into `http_wasm_fetch_stream`.

`pull: (controller: ReadableByteStreamControllerExtension, desired_size: number, pull_state: any) => void`

campersau committed:
rename pull to pull_delegate

campersau committed:
move reading into ReadableStreamPullState

campersau committed:
pass HttpCompletionOption.ResponseHeadersRead in streaming test

pavelsavara committed:
- don't pass controller to C#, only pass pull_state

- move pull and fetch promise controllers to JS pull_state proxy
- removed desiredBytes for simplicity, we could put it back ...
- made it more clear what is async and what is not
- http_wasm_fetch_stream needs t return cancelable promise, not regular promise
- few asserts
- added C# proxy disposal

pavelsavara commented:
I made following changes:

- don't pass controller to C#, only pass pull_state
- move pull and fetch promise controllers to JS pull_state proxy
- removed desiredBytes for simplicity, we could put it back ...
- made it more clear what is async and what is not
- http_wasm_fetch_stream needs t return cancelable promise, not regular promise
- few asserts
- added C# proxy disposal

does it make sense ? Did I break anything ?

campersau commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -57,54 +58,81 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+
+    const body = new ReadableStream({
+        type: "bytes",
+        autoAllocateChunkSize: 65536,
```
This can be removed when we don't use `byob`.

pavelsavara commented:
oh `/p:Outerloop=true`

campersau commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -57,54 +58,81 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
     });
+
+    const cancelable_promise = http_wasm_fetch(url, header_names, header_values, option_names, option_values, abort_controller, body);
+    pull_state.__fetch_promise_control = loaderHelpers.getPromiseController(cancelable_promise);
```
That is nice!

campersau reviewed (commented):
The tests run into `TimeoutExceptions`, currently not sure why.
Yes, these are all outerloop tests :)

pavelsavara commented:
> The tests run into `TimeoutExceptions`, currently not sure why. Yes, these are all outerloop tests :)

I'm debugging it now and will get back. The problem with outerloop tests is that I don't know who owns `corefx-net-http2.azurewebsites.net` and so I don't know how to deploy the new `/echobody.ashx`
Also that beast is not open to CORS. The solution is to use local server as part of the xharness. But somehow, it doesn't work for me right now

Stay tuned.

campersau commented:
> The problem with outerloop tests is that I don't know who owns corefx-net-http2.azurewebsites.net and so I don't know how to deploy the new /echobody.ashx

I thought this is just a local aspnet core kestrel server and not actually an external website.

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -367,30 +351,27 @@ public ReadableStreamPullState(Stream stream, CancellationToken cancellationToke
-                else
-                {
-                    view = _buffer.AsMemory(0, desiredSize);
+                    ReadableStreamControllerEnqueueUnsafe(this, handle, view.Length);
```
`view.Length` should be just `length` now.

campersau replied:
Maybe we could remove the `Memory<byte>` now and pin the byte array manually like here:

https://github.com/dotnet/runtime/blob/76485443684db1e379fc73184d1e2549b9245db3/src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpInterop.cs#L97-L103

pavelsavara commented:
> > The problem with outerloop tests is that I don't know who owns corefx-net-http2.azurewebsites.net and so I don't know how to deploy the new /echobody.ashx
> 
> I thought this is just a local aspnet core kestrel server and not actually an external website.

When configured properly and for wasm, yes. I had broken local environment.
For non-wasm HTTP test, they use real remote call.

campersau committed:
fix build by passing the correct length
- made buffer readonly
- removed autoAllocateChunkSize

pavelsavara committed:
feedback

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -367,30 +351,27 @@ public ReadableStreamPullState(Stream stream, CancellationToken cancellationToke
         }
 
-        public async Task<Memory<byte>> ReadAsync(int desiredSize)
+        public async Task Pull()
```
done

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -354,24 +354,25 @@ public ReadableStreamPullState(Stream stream, CancellationToken cancellationToke
-        private static unsafe void ReadableStreamControllerEnqueueUnsafe(object pullState, Buffers.MemoryHandle handle, int length) =>
-            BrowserHttpInterop.ReadableStreamControllerEnqueue(pullState, (IntPtr)handle.Pointer, length);
+
+        private unsafe void ReadableStreamControllerEnqueueUnsafe(object pullState, int length)
```
Looks like this method isn't getting called?

pavelsavara replied:
![image](https://github.com/dotnet/runtime/assets/271576/1b1c06c1-c670-421e-b3b3-c595ece48624)

pavelsavara replied:
because `type: "bytes"` need to be there.
Now enqueue is not failing, but `http_wasm_readable_stream_controller_enqueue` is never called with zero

campersau replied:
For me it is working with this fix https://github.com/dotnet/runtime/pull/91295/commits/dd83582568b482dd5dda7da0899ce1181fb9e1bd . Depending on which tests you run it might take a while to get to `0` as the send a lot of data.

pavelsavara committed:
type: "bytes" need be there

campersau committed:
call ReadableStreamControllerEnqueueUnsafe again

pavelsavara commented on src/libraries/Common/tests/System/Net/Http/ResponseStreamTest.cs:
```diff
@@ -229,9 +229,148 @@ await client.GetAsync(remoteServer.EchoUri, HttpCompletionOption.ResponseHeaders
 
 #if NETCOREAPP
+
         [OuterLoop]
```
Let's make this one not-OuterLoop

pavelsavara committed:
BrowserHttpHandler_Streaming not OuterLoop

pavelsavara commented:
I like it as it is now, @maraf could you please review it ?

pavelsavara commented:
/azp run runtime-wasm

maraf commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -16,6 +17,22 @@ function verifyEnvironment() {
+export function http_wasm_supports_streaming_request(): boolean {
+    if (typeof Request !== "undefined" && "body" in Request.prototype && typeof ReadableStream === "function") {
+        let duplexAccessed = false;
+        const hasContentType = new Request("", {
```
Could we add here a comment explaining the logic behind feature detection? Something similar to what is in the referenced article.

campersau replied:
Done.

maraf reviewed (approved):
Great job!

campersau committed:
add comment for streaming request feature detection logic

pavelsavara commented:
I would like to see `BrowserHttpHandler_Streaming` running on CI, but I can't find it in the list yet.
I would also expect it to fail on NodeJS, so I think it probably didn't run ?

campersau commented:
> I would also expect it to fail on NodeJS, so I think it probably didn't run ?

The `BrowserHttpHandler_*` tests only run on `PlatformDetection.IsBrowser` which I think does not include NodeJS, right?
Does `TestPlatforms.Browser` include NodeJS? Then I guess `PostAsync_CancelDuringRequestContentSend_TaskCanceledQuickly` should have run and failed, because we don't support streaming there yet?

pavelsavara commented:
`PlatformDetection.IsBrowser` means "browser OS"  and Nodejs is consuming that runtime flavor.

We have plenty of things similar to 
`[ActiveIssue("https://github.com/dotnet/runtime/issues/86317", typeof(PlatformDetection), nameof(PlatformDetection.IsNodeJS))]` which do work on actual browser.

pavelsavara commented:
This is chrome
[[PASS] System.Net.Http.Functional.Tests.SocketsHttpHandler_ResponseStreamTest.BrowserHttpHandler_Streaming](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-91295-merge-5ad59649412e41c789/WasmTestOnBrowser-System.Net.Http.Functional.Tests/1/console.62e69c44.log?helixlogtype=result)

pavelsavara commented:
This is [NodeJS](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-91295-merge-73d89b296dc14ff885/WasmTestOnNodeJS-System.Net.Http.Functional.Tests/1/console.51caafa6.log?helixlogtype=result)

It doesn't run the test because https://github.com/dotnet/runtime/blob/f4702002c277ed5d1fda3cb84cc8a7e68ce231f2/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs#L643

Which is OK

guardrex commented:
@pavelsavara ... The issue and PR are marked for 9.0 milestone. I just want to confirm that this will indeed be for 9.0.

pavelsavara commented:
I didn't plan to backport this to Net8, it's too late for that I think.

@lewing  your expectations ?

pavelsavara commented:
> a better solution would probably be to change the way `MultipartContent.CreateContentReadStreamAsyncCore` works to avoid the buffering.

@greenEkatherine any thought on the feedback above ? 
Should we create new issue for that ? 
Does the proposed fix make sense for all platforms or we should seek WASM specific fix ?

campersau commented:
We could change the implementation to use `HttpContent.CopyToAsync` instead and use a `TransformStream` in JS which could look something like this: https://github.com/campersau/runtime/commit/963f9ff92e1ba1beb30278c00094a9c426002315
